apps/meteor/app/lib/server/functions/saveCustomFieldsWithoutValidation.ts

```diff
@@ -1,49 +1,66 @@
-	if (trim(settings.get('Accounts_CustomFields')) !== '') {
-		let customFieldsMeta;
-		try {
-			customFieldsMeta = JSON.parse(settings.get('Accounts_CustomFields'));
-		} catch (e) {
-			throw new Meteor.Error('error-invalid-customfield-json', 'Invalid JSON for Custom Fields');
-		}
+const getCustomFieldsMeta = function (customFieldsMeta: string) {
+	try {
+		return JSON.parse(customFieldsMeta);
+	} catch (e) {
+		throw new Meteor.Error('error-invalid-customfield-json', 'Invalid JSON for Custom Fields');
+	}
+};
+export const saveCustomFieldsWithoutValidation = async function (
+	userId: string,
+	formData: Record<string, any>,
+	_updater?: Updater<IUser>,
+): Promise<void> {
+	const customFieldsSetting = settings.get<string>('Accounts_CustomFields');
+	if (!customFieldsSetting || trim(customFieldsSetting).length === 0) {
+		return;
+	}
+
+	// configured custom fields in setting
+	const customFieldsMeta = getCustomFieldsMeta(customFieldsSetting);
+
+	const customFields: Record<string, any> = Object.keys(customFieldsMeta).reduce(
+		(acc, currentValue) => ({
+			...acc,
+			[currentValue]: formData[currentValue],
+		}),
+		{},
+	);
 
-		const customFields: Record<string, any> = {};
-		Object.keys(customFieldsMeta).forEach((key) => {
-			customFields[key] = formData[key];
-		});
-		await Users.setCustomFields(userId, customFields);
+	const updater = _updater || Users.getUpdater();
 
-		// Update customFields of all Direct Messages' Rooms for userId
-		const setCustomFieldsResponse = await Subscriptions.setCustomFieldsDirectMessagesByUserId(userId, customFields);
-		if (setCustomFieldsResponse.modifiedCount) {
-			void notifyOnSubscriptionChangedByUserIdAndRoomType(userId, 'd');
+	updater.set('customFields', customFields);
+
+	// add modified records to updater
+	Object.keys(customFields).forEach((fieldName) => {
+		if (!customFieldsMeta[fieldName].modifyRecordField) {
+			return;
 		}
 
-		for await (const fieldName of Object.keys(customFields)) {
-			if (!customFieldsMeta[fieldName].modifyRecordField) {
-				return;
-			}
-
-			const { modifyRecordField } = customFieldsMeta[fieldName];
-			const update: DeepWritable<UpdateFilter<IUser>> = {};
-			if (modifyRecordField.array) {
-				update.$addToSet = {};
-				update.$addToSet[modifyRecordField.field] = customFields[fieldName];
-			} else {
-				update.$set = {};
-				update.$set[modifyRecordField.field] = customFields[fieldName];
-			}
-
-			await Users.updateOne({ _id: userId }, update);
+		const { modifyRecordField } = customFieldsMeta[fieldName];
+
+		if (modifyRecordField.array) {
+			updater.addToSet(modifyRecordField.field, customFields[fieldName]);
+		} else {
+			updater.set(modifyRecordField.field, customFields[fieldName]);
 		}
+	});
+
+	if (!_updater) {
+		await Users.updateFromUpdater({ _id: userId }, updater);
+	}
+
+	// Update customFields of all Direct Messages' Rooms for userId
```

Comment:
Here's a problem: if the updater is received by param, it'll only commit the changes to the user later, but the subscriptions are already being updated here regardless.

If there's any problem with the user update, the user and its subscriptions will be holding different information.